Add comments to the code

khalillechelt opened the pull request:
Hey @staltz there were a few things I did not quite get in the code of the todomvc example. 
In order to understand it better I went ahead and commented the code as I understood it. 
It really helped me to understand what was happening with the `proxyItemAction$` for instance. 

I think this could be useful for other Cycle beginners as well. I am sure it could do with some reviewing and editing. 

So here it is as a proposition in case you would like to use it / integrate it etc.

Commit:
add me as contributor

Commit:
remove old localstorage code

Commit:
add comments

Commit:
edit comment

Commit:
add comments

Commit:
add more comments

staltz commented on src/app.js:
```diff
@@ -1,18 +1,30 @@
+// THE ENTRY POINT
+// This is where the whole story starts.
+// `run` receives a main function and an object with
+// the source streams.
```
"and an object with the source streams"
That's wrong, it's an object with the drivers (each driver is a function). Drivers may generate source streams, but not necessarily. Some drivers just read sinks and consume them, generating no sources at all.

staltz commented:
Good initiative! I'll review the comments.

staltz commented on src/app.js:
```diff
@@ -1,18 +1,30 @@
 run(main, {
+  // THE DOM DRIVER
+  // We create a DOM driver by passing
+  // a DOM selector to the `makeDOMDriver` function.
```
The comment is just paraphrasing the code, so it's not useful. Either remove the comment or explain the purpose of passing a selector to makeDOMDriver.

staltz commented on src/app.js:
```diff
@@ -1,18 +1,30 @@
-  // localStorageSource: CustomDrivers.makeLocalStorageSourceDriver('todos-cycle'),
-  // localStorageSink: CustomDrivers.makeLocalStorageSinkDriver('todos-cycle'),
+  // THE INITAL HASH STREAM
+  // A stream that only delivers the initial hash value.
```
"A **driver** that only delivers the initial hash value as source"

staltz commented on src/app.js:
```diff
@@ -1,18 +1,30 @@
+  // A stream that only delivers the initial hash value.
   initialHash: () => Observable.just(window.location.hash),
+  // THE HASH CHANGE STREAM
+  // A stream that delivers the hash value on the hashChange event.
```
"A **driver** that delivers the hash value on the hashChange event, as source"

staltz commented on src/components/TodoItem.js:
```diff
@@ -2,27 +2,41 @@ import {Observable} from 'rx';
   const sanitizedProps$ = props$.startWith({title: '', completed: false});
+  // THE EDITING STREAM
+  // Create an editing stream that consists only of
+  // actions related to editing.
```
Not exactly what it means. Because of the startWith(false), this stream doesn't have actions (actions are objects with the type and values... like in Redux). This stream emits booleans. It is a stream representing the "is editing" state.

staltz commented on src/components/Todos/index.js:
```diff
@@ -7,17 +7,41 @@ import deserialize from './storage-source';
 import TodoItem from '../TodoItem';
 
+// AMEND STATE WITH CHILDREN
+// This is a callback function for a map function used below.
```
Not really. This function **creates** the callback (I wouldn't call it callback either, it's just a "projection" function) for the map function below.

staltz commented on src/components/Todos/index.js:
```diff
@@ -7,17 +7,41 @@ import deserialize from './storage-source';
             DOM: todoItem.DOM,
+            // The action stream for the todo item is enhanced with
+            // the id of the todo, in order to tell which item these actions
+            // are streaming from.
```
I wouldn't merge any of these comments above in this function, because they are just paraphrasing the code. These types of comments don't really add value, they just re-explain the code, but add a lot of visual pollution. I would only add one line of comment to explain the **purpose** of this function: why it exists.

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
 
+// THE MAIN FUNCTION
+// This is the implementation of the `main` function that gets passed to
+// Cycle's `run` function.
```
We should not assume from this file that this is the `main` given to Cycle.run. That's a decision made only in src/app.js. As far as this file is concerned, it only exports the `Todos` component. This file itself is not opinionated that this component should be the top-level one. That's the point of Cycle Nested: nothing stops you from reusing this `Todos` component inside a larger app.

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
+  // THE LOCALSTORAGE STREAM
+  // Here we create a localStorage stream that only streams
+  // the first value received from localStorage because we use
+  // the localStorage to "rehydrate" the app whenever we launch or refresh it.
```
"because" :+1:
"rehydrate" :-1: because rehydrate is a specific term used to refer to isomorphic rendering where the state is bundled, serialized, passed to the client, then "hydrated" to become live state. This here is just reading stored initial data.

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
   const sourceTodosData$ = deserialize(localStorage$);
+  // THE PROXY ITEM ACTION STREAM
+  // We use an Rx.Subject as a proxy for all the actions that stream
+  // from the different todo items.
```
:+1:

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
   const proxyItemAction$ = new Subject();
+  // THE INTENT (MVI PATTERN)
+  // Pass relevant sources to the intent funtion, which set up
+  // streams that model the users intentions.
```
:+1:

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
+  // streams that model the users intentions.
   const actions = intent(DOM, hashchange, initialHash, proxyItemAction$);
+  // THE MODEL (MVI PATTERN)
+  // Intent gets passed to the model function which transforms the data
```
Actions get passed to the model, not Intent.

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +50,56 @@ function amendStateWithChildren(DOMSource) {
+  // coming through the intent streams and prepares the data for the view.
   const state$ = model(actions, sourceTodosData$);
+  // AMEND STATE WITH CHILDREN
+  // The state is being amended with the children.
```
This is just a paraphrasing comment.

staltz commented on src/components/Todos/intent.js:
```diff
@@ -1,31 +1,52 @@
   return {
+    // THE ROUTE STREAM
+    // The stream of the initial hash is being concatenated with
+    // the hashchange stream.
```
Paraphrasing. Explaining purpose would be better.

staltz commented on src/components/Todos/intent.js:
```diff
@@ -1,31 +1,52 @@
     editTodo$: itemAction$.filter(action => action.type === 'doneEdit'),
 
+    // TOGGLE ALL STREAM
+    // Create a stream out of the clicks on the `.toggle-all` button.
```
:+1:

staltz commented on src/components/Todos/model.js:
```diff
@@ -1,5 +1,7 @@
 import {Observable} from 'rx';
 
+// A helper function to get filter function to filter the todo items
+// in the virtual dome according to which route we are on.
```
This comment is really confusing. Also...
`dome` typo

staltz commented on src/components/Todos/model.js:
```diff
@@ -8,6 +10,7 @@ function getFilterFn(route) {
   }
 }
 
+// A helper to search for the todo index.
```
Paraphrasing. Explaining purpose would be better.

staltz commented on src/components/Todos/model.js:
```diff
@@ -27,6 +30,10 @@ function searchTodoIndex(todosList, todoid) {
 }
 
+// MAKE MODIFICATION STREAM
+// A function that takes the intent on the todo list
```
intent is the function. actions are the output of the intent. This function here takes actions, not intent.

staltz commented on src/components/Todos/model.js:
```diff
@@ -27,6 +30,10 @@ function searchTodoIndex(todosList, todoid) {
+// MAKE MODIFICATION STREAM
+// A function that takes the intent on the todo list
+// and returns a stream of functions that expect todosData (the data model)
+// and return a modified version of the data.
```
This comment could explain what those functions are. They are "updater" functions. Sort of like reducers in Redux, they take old state and return new state.

staltz commented on src/components/Todos/model.js:
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
+  // THE BUSINESS LOGIC
+  // Actions are passed to the `makeModification$` function
+  // which creates a stream of modification functions that needs
+  // to be applied on the todoData when an action happens.
```
:+1:

staltz commented on src/components/Todos/model.js:
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
     .concat(modification$)
+    // ...and apply every modification function on the
+    // current todosData which is the "accumulater"
+    // of this `.scan` function.
```
These comments above are too much paraphrasing.

staltz commented on src/components/Todos/model.js:
```diff
@@ -93,12 +100,28 @@ function makeModification$(actions) {
+    // of this `.scan` function.
     .scan((todosData, modFn) => modFn(todosData))
+    // Always serve any subscriber the latest state that
+    // was flowing through this Observable.
```
`share` makes it hot, which means it multicasts/broadcasts the Observable for many observers.

staltz commented on src/components/Todos/view.js:
```diff
@@ -15,6 +15,8 @@ function renderHeader() {
 
 function renderMainSection(todosData) {
+
+  console.log(todosData.filterFn.toString());
```
Could we not have console log?

staltz commented on src/components/Todos/view.js:
```diff
@@ -24,6 +26,8 @@ function renderMainSection(todosData) {
     }),
     ul('.todo-list', todosData.list
+      // Apply filter function according to
+      // route. 
```
Correct, and this is not paraphrasing, **but** from the view we should make no assumption that it is "according to the route". The model is the only one who knows that information, the view should be agnostic. All the view knows is that there may be some filtering logic, and it is going to apply that.

staltz commented on src/components/Todos/view.js:
```diff
@@ -69,6 +73,11 @@ function renderFooter(todosData) {
+// This function expects the stream of todosData
+// from the model function and turns it into a
+// virtual DOM stream that is then ultimately returned intot
+// the DOM sink in the index.js.
```
:+1:
Except a typo `intot`

khalillechelt commented:
Thanks a million for the corrections! Will get on the fixes this weekend.

Commit:
comment fix

This addresses line note https://github.com/cyclejs/todomvc-cycle/pull/21/files#r49221502 by @staltz.

Commit:
comment fix

This addresses line note https://github.com/cyclejs/todomvc-cycle/pull/21/files#r49221611.

Commit:
comment fix

Commit:
add various comment fixes

khalillechelt commented:
@staltz OK, I changed all the bits that you pointed out.

khalillechelt commented:
I am wondering if it would make sense to turn these comments into some sort of step-through guide that allows a reader to step through a complete "cycle" maybe starting off from a user's action to the updated DOM. What do you think?

staltz commented on src/components/Todos/index.js:
```diff
@@ -7,15 +7,27 @@ import deserialize from './storage-source';
       list: todosData.list.map(data => {
+        // Turn the data item into an Observable
         const props$ = Observable.just(data);
+        // Create scoped todo item data flow components.
```
Would be good if `data flow` is `dataflow`, and this line should say "component" not "components".

staltz commented on src/components/Todos/index.js:
```diff
@@ -7,15 +7,27 @@ import deserialize from './storage-source';
+        // Create scoped todo item data flow components.
         const todoItem = isolate(TodoItem)({DOM: DOMSource, props$});
+
+        // Return the new data item for the list property arrayj.
```
Typo "arrayj"

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +38,54 @@ function amendStateWithChildren(DOMSource) {
+// THE TODOS FUNCTION
+// This is the Todos component which is being exported below.
+// The sources that have been passed to `run` as well are immediately
+// deconstructed into the variables `DOM`, `hashchange`, `initialHash` & `storage`.
```
I would rephrase this sentence as: "Using destructuring, we pick the sources `DOM`, `hashchange`, `initialHash` and `storage` from the sources object as argument."

staltz commented on src/components/Todos/index.js:
```diff
@@ -26,20 +38,54 @@ function amendStateWithChildren(DOMSource) {
 function Todos({DOM, hashchange, initialHash, storage}) {
+  // THE LOCALSTORAGE STREAM
+  // Here we create a localStorage stream that only streams
+  // the first value received from localStorage in order to
```
I would replace `received` with `read` here in order to emphasize "sources" == "reads".

staltz commented:
@kahlil almost perfect, just a few minor issues.

Commit:
comment fixes

khalillechelt commented:
@staltz thanks! Fixed.

staltz commented:
Alright, merging. Thanks a lot!
